app/assets/javascripts/discourse/app/static/prosemirror/extensions/mention.js

```diff
@@ -74,6 +61,150 @@ const extension = {
       }
     },
   },
+
+  plugins({ pmState: { Plugin, PluginKey } }) {
+    const plugin = new PluginKey("mention");
+
+    return new Plugin({
+      key: plugin,
+      view() {
+        return {
+          update(view) {
+            if (!view._existingMentionsValidated) {
+              this.processExistingMentions(view);
+              view._existingMentionsValidated = true;
+            }
+
+            this.processNewMentions(view);
+          },
+          processExistingMentions(view) {
+            const nodeList = [];
+
+            view.state.doc.descendants((node, pos) => {
+              if (node.type.name === "mention") {
+                nodeList.push({ node, pos });
+                pendingMentions.add(node.attrs.name);
+              }
+            });
+
+            if (!nodeList.length) {
+              return;
+            }
+
+            // process in reverse to avoid issues with position shifts
+            nodeList.sort((a, b) => b.pos - a.pos);
+
+            const invalidateMentions = async () => {
+              await fetchMentions([...pendingMentions]);
+
+              for (const item of nodeList) {
+                const { node, pos } = item;
+                const name = node.attrs.name;
+                const isValid = validMentions.has(name);
+
+                if (!isValid) {
+                  view.dispatch(
+                    view.state.tr
+                      .delete(pos, pos + node.nodeSize)
+                      .insertText(`@${name}`, pos)
+                  );
+                }
+              }
+            };
+
+            invalidateMentions();
+          },
+          processNewMentions(view) {
+            const tr = view.state.tr;
+
+            if (!isBoundary(tr.doc.textContent.slice(-1), 0)) {
+              return;
+            }
+
+            view.state.doc.descendants((node, pos) => {
+              if (node.type.name === "text") {
```

Comment:
To remove a level of indent here, can you do `if (node.type.name !== "text") { next }` or something here?